### PR TITLE
[core] Isolate classloaders for runtime and auxclasspath

### PR DESCRIPTION
**Pull request by jsotuyod**
- We will now always load classes from auxclasspath during
typeresolution, even if we have a class with the same name in PMD's
classpath
 - This prevents conflicts when using different versions of the same
dependencies

### CODE REVIEW

File: pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactory.java [hunk 1]
```diff
@@ -543,7 +543,7 @@ private void parseSingleRuleNode(RuleSetReferenceId ruleSetReferenceId, RuleSetB
         if (attribute == null || "".equals(attribute)) {
             throw new IllegalArgumentException("The 'class' field of rule can't be null, nor empty.");
         }
-        Rule rule = (Rule) classLoader.loadClass(attribute).newInstance();
+        Rule rule = (Rule) RuleSetFactory.class.getClassLoader().loadClass(attribute).newInstance();
```
Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactory.java]: This change in particular means users can no longer write custom Java rules within their project's codebase / auxclasspath and apply them. This feature was undocumented and unsupported.

Custom rules can be written, but must be part of PMD's execution classpath.

**Comment by adangel** [comment]
@jsotuyod Is this a fix for the build problem described in #679?

> Custom rules can be written, but must be part of PMD's execution classpath.

This sounds totally correct - I wouldn't have expected to be implemented otherwise. But it was... With this PR, the classes need to be on the execution classpath, but the rulesets still can be on the auxclasspath, which is wrong, I think.

I don't know, why we use the auxclasspath in the RuleSetFactory at all. It looks like, it was a wrong fix some time ago, for the problem, that PMD didn't find a rule/ruleset (because it's easier to change the auxclasspath -> command line option, than to change the execution class path -> shell script).

One point to the classloading: The maven plugin configures the Thread's context classloader - which might make a difference - not sure.

I'm not  sure, whether we really need the change for the parent-last strategy. It would only make a difference, if the same class is multiple times on the auxclasspath...

**Comment by jsotuyod** [comment]
@adangel yes, this addresses the issue on #679, I was finally able to reproduce it even from outside Maven.

You are right, the ruleset can be in the auxclasspath. I believe the reason why we give a classloader to the rule set factory is here:

https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/ant/internal/PMDTaskImpl.java#L240-L253

It can probably be worked out more elegantly.

> I'm not sure, whether we really need the change for the parent-last strategy. It would only make a difference, if the same class is multiple times on the auxclasspath...

If a project uses any dependency we use, but a different version, they will have a conflict. Suppose someone is using `commons-io` 2.6. Since we use version 2.4, and the default is parent-first, we will always try to load classes from the execution claspath before checking the auxclasspath. Therefore, all classes will be loaded from 2.4, meaning typeresolution can fail (incompatible types, missing methods, etc.). We want to use 2.4 for running PMD, but 2.6 for typeresolution.

**Comment by adangel** [comment]
@jsotuyod Ok, understood the issue why parent-last.

But: Should the execution classpath ever end up in the auxclasspath, like it currently always does?  https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/PMDConfiguration.java#L191-L196 (either classpath is null or set via the method you mentioned for ant)

We probably should only have on the auxclasspath, whatever is given as the auxclasspath configuration option + the standard java base classpath (java.lang.*, etc.) / bootclasspath.

**Comment by jsotuyod** [comment]
> We probably should only have on the auxclasspath, whatever is given as the auxclasspath configuration option + the standard java base classpath (java.lang.*, etc.) / bootclasspath.

@adangel I thought about it, but I don't know if we can get a hold of the same "boot classloader" as the one being used to run PMD. Otherwise I see 2 issues:

1. we would end up loading twice all JRE classes, meaning more memory / CPU usage
2. rules would be harder to write. We can't check against stuff such as `instanceof String` anymore, as that `String` would be loaded from a different classloader than the one store in the node's type, and therefore not match.

Code such as this will fail the exact same way:

https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symboltable/TypeSet.java#L402-L409

I'm uncertain if setting the parent classloader to [`getSystemClassLoader()`](https://docs.oracle.com/javase/7/docs/api/java/lang/ClassLoader.html#getSystemClassLoader()) will work for this. Wouldn't this produce the same behavior we currently have when running from CLI (java -cp allpmdclassesanddeps)?

File: pmd-core/src/main/java/net/sourceforge/pmd/util/ClasspathClassLoader.java [hunk 2]
```diff
@@ -96,4 +96,27 @@ public String toString() {
                 .append(StringUtils.join(getURLs(), ":"))
                 .append("] parent: ").append(getParent()).append(']').toString();
     }
+    
+    @Override
+    protected Class<?> loadClass(final String name, final boolean resolve) throws ClassNotFoundException {
+        // First, check if the class has already been loaded
+        Class<?> c = findLoadedClass(name);
+        if (c == null) {
+            if (c == null) {
```
Comment (gibarsin) [inline pmd-core/src/main/java/net/sourceforge/pmd/util/ClasspathClassLoader.java]: Nothing harmful, but the null check is done twice.

Comment (jsotuyod) [reply pmd-core/src/main/java/net/sourceforge/pmd/util/ClasspathClassLoader.java]: Thanks! Bad copy-paste on my part

**Comment by jsotuyod** [comment]
@adangel I confirm, [`getSystemClassLoader()`](https://docs.oracle.com/javase/7/docs/api/java/lang/ClassLoader.html#getSystemClassLoader()) returns the same classloader used to start the app, and contains the whole classpath.

We could, request that classloader's parent, and that one DOES only resolve JRE classes, but I am not sure how reliable that is... I can't find documentation saying this behavior is standard.

**Comment by jsotuyod** [comment]
@adangel I reworked how we load rulesets. In doing so we now enforce both rules and rulesets are only in the execution classpath.

As for typeresolution, we keep the newly introduced parent-last approach to avoid issues.

I'm no longer adding `classpath` entries defined in Ant to the auxclasspath (although they can still be used to locate rulesets). Users should add those entries to the auxclasspath explicitly if they want them.

Please review it, and provide feedback on the open issue on which parent classloader to use for typeresolution.

**Comment by adangel** [comment]
> ...  We can't check against stuff such as instanceof String anymore

Completely agree - this would be a no-go.

> We could, request that classloader's parent, and that one DOES only resolve JRE classes, but I am not sure how reliable that is... I can't find documentation saying this behavior is standard.

Probably not reliable :) At least, the documentation about the `getSystemClassloader` talks about implementation-dependency... 

So, let's stay with the parent-last approach for resolving classing during typeresolution.

I like, how you reworked ruleset loading with using the resource loader! Nice!

**Pull request merged**